michbl/assignment3/Dictionary.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef DICTIONARY_HPP
+#define DICTIONARY_HPP
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <unordered_map>
+#include <set>
+#include "Grid.hpp"
+
+#define END_OF_WORD -1
+
+using namespace std;
+
+// implementing dictionary as trie
+class Dictionary {
+private:
+    struct Node {
+        char c;
+        unordered_map<char, Node *> children;
+
+        Node(char c) : c(c) {
+            this->children.clear();
+        }
+    };
+
+    Node *root;
+    vector<Node *> nodes; // for memory release
+
+    void wordSearchBT(Grid &grid, int row, int col, set<string> &res, string word) {
+        char c;
+        if (!grid.getChar(row, col, c)) {
+            return;
+        }
+        word += c;
+        if (isWord(word)) {
+            res.insert(word);
+        }
+        if (isPrefix(word)) {
+            Direction dirs[8] = {LEFT, RIGHT, UP, DOWN, TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT};
+            int nextRow, nextCol;
+            for (Direction dir : dirs) {
+                grid.getNextIdxs(row, col, nextRow, nextCol, dir);
+                wordSearchBT(grid, nextRow, nextCol, res, word);
+            }
+        }
+    }
+
+public:
+    Dictionary() {
+        this->root = new Node(END_OF_WORD);
+        nodes.push_back(this->root);
+    }
+
+    Dictionary(const vector<string> words) {
+        this->root = new Node(END_OF_WORD);
+        nodes.push_back(this->root);
+        for (string word : words) {
+            insertWord(word);
+        }
+    }
+
+    ~Dictionary() {
+        for (auto node : this->nodes) {
+            delete node;
+        }
+    }
+
+    void insertWord(const string &word) {
+        unordered_map<char, Node *> *currentChildren = &(this->root->children);
+        unordered_map<char, Node *>::iterator it;

```

Comment:
The benefit of defining a variable before the `for` loop is efficiency - instead of creating the same variable over and over, it is created once and then overridden.
The downside is readability - it is harder to understand the context of this iterator. The following is more readable in my opinion:
```C++
for (char ch : sord) {
    unordered_map<char, Node *>::iterator it = currentChildren->find(ch);
    // Can also be:    auto it = currentChildren->find(ch);
    if (it != ...
   ...
}
```
but can be less efficient, especially if the repeating variable is heavy to construct or allocate.

Both are OK, and it is good to consider efficiency when writing code. It is up to you to decide which fits in each case. Just remember:
> *"Premature optimization is the root of all evil" - Donald Knuth*

;)

Reply:
I'll keep that in mind from now on :)